File: proposed-projects/round-trip-experiment/reverse/20260909T020947Z_ffb57550/20260910T054605Z_51904c27/ported_implementation/gbnf/GBNF.py
from .grammar_graph.graph import Graph
from .grammar_graph.parse_state import ParseState
from .grammar_parser.build_rule_stack import build_rule_stack
from .rules_builder.rules_builder import RulesBuilder
from .utils.errors import GrammarParseError


def GBNF(grammar: str, initial_string: str = "") -> ParseState:
    if not isinstance(grammar, str):
        raise TypeError("grammar must be a string")

    if not isinstance(initial_string, str):
        raise TypeError("input must be a string")

    builder = RulesBuilder(grammar)
    rules, symbol_ids = builder.rules, builder.symbol_ids
    if len(rules) == 0:
        raise GrammarParseError(grammar, 0, "No rules were found")
    root_id = symbol_ids.get("root")
    if root_id is None:
        raise GrammarParseError(
            grammar,
            0,
            "Grammar does not contain a 'root' symbol",
        )

    stacked_rules = [build_rule_stack(rule) for rule in rules]
    graph = Graph(grammar, stacked_rules, root_id)
    return ParseState(graph, graph.add(initial_string))

File: packages/gbnf-experiment/data/20260909T004354Z_a47a3057/ported_implementation/src/gbnf.ts
import { Graph } from './grammar-graph/graph.js';
import { ParseState } from './grammar-graph/parse-state.js';
import { buildRuleStack } from './grammar-parser/build-rule-stack.js';
import { RulesBuilder } from './rules-builder/rules-builder.js';
import { GrammarParseError } from './utils/errors/index.js';

export const GBNF = (grammar: string, initialString = ''): ParseState => {
  if (typeof grammar !== 'string') {
    throw new Error('grammar must be a string');
  }

  if (typeof initialString !== 'string') {
    throw new Error('input must be a string');
  }

  const rulesBuilder = new RulesBuilder(grammar);
  const { rules, symbolIds } = rulesBuilder;
  if (rules.length === 0) {
    throw new GrammarParseError(grammar, 0, 'No rules were found');
  }
  const rootId = symbolIds.get('root');
  if (rootId === undefined) {
    throw new GrammarParseError(
      grammar,
      0,
      "Grammar does not contain a 'root' symbol",
    );
  }

  const stackedRules = rules.map((rule) => buildRuleStack(rule));
  const graph = new Graph(grammar, stackedRules, rootId);
  return new ParseState(graph, graph.add(initialString));
};

export default GBNF;
